# 06 - Raven calibration

## Calibration of a Raven model

In this notebook, we show how to calibrate a Raven model using the GR4J-CN predefined structure. The users can refer to the documentation for the parameterization of other hydrological model structures.

Let's start by importing the packages that will do the work.

- ravenpy.models.GR4JCN: The Raven GR4JCN model Python wrapper we used in the 2 previous notebooks
- ravenpy.models.GR4JCN_OST: The Raven GR4JCN model OSTRICH optimization python wrapper. It basically allows us to easily calibrate a Raven model.

Depending on your model choice (HMETS, HBVEC, etc.), simply add "\_OST" to the end to get the OSTRICH wrapper. 

In [1]:
import tempfile
import datetime as dt
import spotpy

from ravenpy.new_config import commands as rc
from ravenpy.new_config.emulators import GR4JCN

from ravenpy.utilities.new_config.calibration import SpotSetup

## Preparing the model to be calibrated on a given watershed
Our test watershed from the last notebook is selected for this test. It can be replaced with any desired watershed.

In [2]:
from ravenpy.utilities.testdata import get_file

# We get the netCDF for testing on a server. You can replace the getfile method by a string containing the path to your own netCDF
nc_files = get_file(
    "raven-gr4j-cemaneige/Salmon-River-Near-Prince-George_meteo_daily.nc"
)

#The path were we saved the results
tempdir = tempfile.mkdtemp()

# Display the datasets that we will be using
print(nc_files)

C:\Users\PC\.raven_testing_data\master\raven-gr4j-cemaneige\Salmon-River-Near-Prince-George_meteo_daily.nc


The process is very similar to setting up a hydrological model. We first need to create the model with its parameters. 

In [3]:
# Here, we need to give the name of your different dataset in order to match with Raven models.
alt_names = {
            "RAINFALL": "rain",
            "TEMP_MIN": "tmin",
            "TEMP_MAX": "tmax",
            "PET": "pet",
            "HYDROGRAPH": "qobs",
            "SNOWFALL": "snow",
}

# The HRU of your watershed
hru = dict(
    area=4250.6, elevation=843.0, latitude=54.4848, longitude=-123.3659
)

# You can decide the evaluation metrics that will be used to calibrate the parameters of your model. You need atleast 
# 1 evaluation metric, but you can do any combinaison of evaluations from this list : 
#
# NASH_SUTCLIFFE,
# LOG_NASH, 
# RMSE, 
# PCT_BIAS, 
# ABSERR, 
# ABSMAX, 
# PDIFF,
# TMVOL,
# RCOEFF,
# NSC,
# KLING_GUPTA
eval_metrics = ("NASH_SUTCLIFFE",)


# We need to create the desired model with its parameters the same way as in the Notebook 04_Emulating_hydrological_models.
model = GR4JCN(
    ObservationData=rc.ObservationData.from_nc(nc_files, alt_names="qobs"),
    Gauge=rc.Gauge.from_nc(
        nc_files,
        alt_names=alt_names,
        extra={1: {"elevation": hru["elevation"]}},
    ),
    HRUs=[hru],
    StartDate=dt.datetime(1990, 1, 1),
    EndDate=dt.datetime(1999, 12, 31),
    RunName="test", 
    EvaluationMetrics=eval_metrics
)

## Spotpy Calibration

Once you've created your model, you need to create a SpotSetup, which will be used to calibrate your model.

In [4]:
# In order to calibrate your model, you need to give the lower and higher bounds of the model. In this case, we are passing
# the boundaries for a GR4JCN, but it's important to change them, if you are using another model. Note that the list of these
# boundaries for each model is at the end of this notebook.
low_params = (0.01, -15.0, 10.0, 0.0, 1.0, 0.0)
high_params = (2.5, 10.0, 700.0, 7.0, 30.0, 1.0)


spot_setup = SpotSetup(
    config=model,
    low=low_params,
    high=high_params,
    path=tempdir,
)

c:\users\pc\documents\github\ravenpy\ravenpy\utilities\new_config\calibration.py:35: UserWarning: Add the `SuppressOutput` command to the configuration to reduce IO.
  warnings.warn(


And then, you need to create a sampler for the SpotSetup which will optimized the paramaters of your model. You can see that we are using the DDS method to optimize the paramaters. If you want to change it to another optimisator, please refer yourself to the Spotpy documentation, here : https://spotpy.readthedocs.io/

Finally, we run the sampler by the amount of desired repetitions.

In [5]:
iterations = 10 # This value should be over 500 for real optimisation. This will take a LONG period of time so be sure of all
                 # the previous setup before executing with a high number.

sampler = spotpy.algorithms.dds(
    spot_setup, dbname="RAVEN_model_run", dbformat="ram", save_sim=False
)

sampler.sample(iterations, trials=1)

Initializing the  Dynamically Dimensioned Search (DDS) algorithm  with  10  repetitions
The objective function will be maximized
Starting the DDS algotrithm with 10 repetitions...
Finding best starting point for trial 1 using 5 random samples.
0
Initialize database...
['csv', 'hdf5', 'ram', 'sql', 'custom', 'noData']


c:\users\pc\documents\github\ravenpy\ravenpy\ravenpy.py:287: RavenWarning: Gauge:Initialize: both snowfall and rainfall data are provided at a gauge, but :RainSnowFraction method is something other than RAINSNOW_DATA. Snow fraction will be recalculated.
  warn(msg, category=RavenWarning)
c:\users\pc\documents\github\ravenpy\ravenpy\ravenpy.py:287: RavenWarning: PET data supplied at gauge Gauge_1 but will not be used due to choice of forcing generation algorithm
  warn(msg, category=RavenWarning)
c:\users\pc\documents\github\ravenpy\ravenpy\ravenpy.py:287: RavenWarning: The required parameter RAIN_ICEPT_PCT for vegetation class VEG_ALL was autogenerated with value 0.120000
  warn(msg, category=RavenWarning)
c:\users\pc\documents\github\ravenpy\ravenpy\ravenpy.py:287: RavenWarning: The required parameter SNOW_ICEPT_PCT for vegetation class VEG_ALL was autogenerated with value 0.100000
  warn(msg, category=RavenWarning)
c:\users\pc\documents\github\ravenpy\ravenpy\ravenpy.py:287: RavenWar

0
2 of 10, maximal objective function=0.140197, time remaining: 00:00:05
0
0
4 of 10, maximal objective function=0.140197, time remaining: 00:00:04
0
0
6 of 10, maximal objective function=0.198817, time remaining: 00:00:03
0
0
8 of 10, maximal objective function=0.198817, time remaining: 00:00:01
0
0
10 of 10, maximal objective function=0.203254, time remaining: 23:59:59
Best solution found has obj function value of 0.203254 at 5



*** Final SPOTPY summary ***
Total Duration: 10.15 seconds
Total Repetitions: 10
Maximal objective value: 0.203254
Corresponding parameter setting:
GR4J_X1: 2.17939
GR4J_X2: 1.91164
GR4J_X3: 543.148
GR4J_X4: 4.90395
CEMANEIGE_X1: 1.26336
CEMANEIGE_X2: 0.7194
******************************



[{'sbest': spotpy.parameter.ParameterSet(),
  'trial_initial': [2.1392328783394814,
   -0.09781184810984023,
   603.0138178282762,
   3.944080569271965,
   10.434355513804867,
   0.4774219467260803],
  'objfunc_val': 0.203254}]

## Analysing the calibration results
The best parameters as well as the objective functions can be analyzed.

In [6]:
# Get the model diagnostics
diag = spot_setup.diagnostics

# Print the NSE and the parameter set in 2 different ways:
print("Nash-Sutcliffe value is: " + str(diag["DIAG_NASH_SUTCLIFFE"]))

# Get all the values of each iteration
results=sampler.getdata() 

#
spotpy.analyser.get_best_parameterset(results) 

Nash-Sutcliffe value is: [0.203254]
Best parameter set:
GR4J_X1=2.179390337556251, GR4J_X2=1.9116435117514765, GR4J_X3=543.1476198470175, GR4J_X4=4.903950072379301, CEMANEIGE_X1=1.263357504341629, CEMANEIGE_X2=0.7194001969526824


array([(2.17939034, 1.91164351, 543.14761985, 4.90395007, 1.2633575, 0.7194002)],
      dtype={'names': ['parGR4J_X1', 'parGR4J_X2', 'parGR4J_X3', 'parGR4J_X4', 'parCEMANEIGE_X1', 'parCEMANEIGE_X2'], 'formats': ['<f8', '<f8', '<f8', '<f8', '<f8', '<f8'], 'offsets': [8, 16, 24, 32, 40, 48], 'itemsize': 64})

## Next steps

In the next notebooks, we will apply the model to specific use-cases, including making and using hotstart files for forecasting, performing hindcasting and forecasting, applying data assimilation and evaluating the impacts of climate change on the hydrology of a watershed. 

## List of Model-Boundaries

GR4JCN : 

<ul>
  <li>low  =  (0.01, -15.0, 10.0, 0.0, 1.0, 0.0),</li>
  <li>high =  (2.5, 10.0, 700.0, 7.0, 30.0, 1.0)</li>
</ul>

         
         
         
HMETS  : 

<ul>
  <li>low  =  (0.3, 0.01, 0.5, 0.15, 0.0, 0.0, -2.0, 0.01, 0.0, 0.01, 0.005, 
               -5.0, 0.0, 0.0, 0.0, 0.0, 0.00001, 0.0, 0.00001, 0.0, 0.0),</li>
  <li>high =  (20.0, 5.0, 13.0, 1.5, 20.0, 20.0, 3.0, 0.2, 0.1, 0.3, 0.1, 
               2.0, 5.0, 1.0, 3.0, 1.0, 0.02, 0.1, 0.01, 0.5, 2.0)</li>
</ul>
                 
                 
Mohyse : 

<ul>
  <li>low  =  (0.01, 0.01, 0.01, -5.00, 0.01, 0.01, 0.01, 0.01, 0.01, 0.01),</li>
  <li>high =  (20.0, 1.0, 20.0, 5.0, 0.5, 1.0, 1.0, 1.0, 15.0, 15.0)</li>
</ul>

         
HBVEC  : 

<ul>
  <li>low  =  (-3.0, 0.0, 0.0, 0.0, 0.0, 0.3, 0.0, 0.0, 0.01, 0.05, 0.01, 
               0.0, 0.0, 0.0, 0.0, 0.0, 0.01, 0.0, 0.05, 0.8, 0.8),</li>
  <li>high =  (3.0, 8.0, 8.0, 0.1, 1.0, 1.0, 7.0, 100.0, 1.0, 0.1, 6.0, 
               5.0, 5.0, 0.2, 1.0, 30.0, 3.0, 2.0, 1.0, 1.5, 1.5)</li>
</ul>
                
                 
CanadianShield :

<ul>
  <li>low  =  (0.01, 0.01, 0.01, 0.0, 0.0, 0.05, 0.0, -5.0, -5.0, 0.5, 0.5, 
               0.0, 0.0, 0.0, 0.0, 0.0, 0.01, 0.005, -3.0, 0.5, 5.0, 0.0, 
               0.0, -1.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.8, 0.8, 0.0),</li>
  <li>high =  (0.5, 2.0, 3.0, 3.0, 0.05, 0.45, 7.0, -1.0, -1.0, 2.0, 2.0,
            100.0, 100.0, 100.0, 0.4, 0.1, 0.3, 0.1, 3.0, 4.0, 500.0, 5.0,
            5.0, 1.0, 8.0, 20.0, 1.5, 0.2, 0.2, 10.0, 10.0, 1.2, 1.2, 1.0)</li>
</ul>

HYPR :

<ul>
  <li>low  =  (-1.0, -3.0, 0.0, 0.3, -1.3, -2.0, 0.0, 0.1, 0.4, 0.0, 0.0,
            0.0, 0.0, 0.0, 0.01, 0.0, 0.0, 1.5, 0.0, 0.0, 0.8),</li>
  <li>high =  (1.0, 3.0, 0.8, 1.0, 0.3, 0.0, 30.0, 0.8, 2.0, 100.0, 
            0.5, 5.0, 1.0, 1000.0, 6.0, 7.0, 8.0, 3.0, 5.0, 5.0, 1.2)</li>
</ul>

SACSMA : 

<ul>
  <li>low  =  (-3.0, -1.52287874, -0.69897, 0.025, 0.01, 0.075, 0.015, 0.04, 
              0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.3, 0.01, 0.8, 0.8),</li>
  <li>high =  (-1.82390874, -0.69897, -0.30102999, 0.125, 0.075, 0.3, 0.3, 0.6, 
              0.5, 3.0, 80.0, 0.8, 0.05, 0.2, 0.1, 0.4, 8.0, 20.0, 5.0, 1.2, 1.2)</li>
</ul>

Blended : 

<ul>
  <li>low  =  (0.0, 0.1, 0.5, -5.0, 0.0, 0.5, 5.0, 0.0, 0.0, 0.0, -5.0,
            0.5, 0.0, 0.01, 0.005, -5.0, 0.0, 0.0, 0.0, 0.3, 0.01, 0.5,
            0.15, 1.5, 0.0, -1.0, 0.01, 0.00001, 0.0, 0.0, -3.0, 0.5,
            0.8, 0.8, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0),</li>
  <li>high =  (1.0, 3.0, 3.0, -1.0, 100.0, 2.0, 10.0, 3.0,
            0.05, 0.45, -2.0, 2.0, 0.1, 0.3, 0.1, 2.0, 1.0,
            5.0, 0.4, 20.0, 5.0, 13.0, 1.5, 3.0, 5.0, 1.0,
            0.2, 0.02, 0.5, 2.0, 3.0, 4.0, 1.2, 1.2, 0.02,
            1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0)</li>
</ul>
